herd/ASLSem.ml

```diff
@@ -242,11 +267,25 @@ module Make (C : Config) = struct
           function
           | V.Val (Constant.Concrete (ASLScalar.S_Bool b)) ->
               V.Val (Constant.Concrete (ASLScalar.S_Bool (not b))) |> return
-          | v -> M.op1 Op.Not v >>= M.op1 (Op.ArchOp1 ASLValue.ASLArchOp.ToBool)
-          )
+          | V.Val (Constant.Symbolic _) as v ->
+              let v' = V.fresh_var () in
+              M.restrict M.VC.[ Assign (v', Unop (Op.Not, v)) ] >>! v'
+          | v -> M.op1 Op.Not v >>= M.op1 (Op.ArchOp1 ASLValue.ToBool))
       | NEG -> M.op Op.Sub V.zero
-      | NOT ->
-          fun v -> M.op1 Op.Inv v >>= M.op1 (Op.ArchOp1 ASLValue.ASLArchOp.ToBV)
+      | NOT -> (
+          function
+          | V.Val (Constant.Symbolic _) as v ->
+              let v' = V.fresh_var () in
+              M.restrict M.VC.[ Assign (v', Unop (Op.Inv, v)) ] >>! v'
+          | v -> M.op1 Op.Inv v >>= to_bv)
```

Comment:
Ah, maybe you want to have an additional constraint in case `v` is a "rigid" symbol (instead of failing as it would be the case in other semantics). In such a case, this looks the right way to proceed. That is, leave most of the job to `M.op1`. If so, why not write `to_bv`, `to_int` etc. in the same style?